### PR TITLE
Add support for Lightweight tags

### LINKED ISSUE #31: Null check operator used on a null value
Hi,
I got the following error getting tags from my local git repository using **git 2.0.0**:

```
Unhandled exception:
Null check operator used on a null value
#0      Tag.parseCatFile (package:git/src/tag.dart:35:40)
#1      GitDir.tags (package:git/src/git_dir.dart:71:17)
```

My code worked fine when the repository had 2 tags but now, with 3 tags my code fails getting the third tag info, running the following code:

```
    final tags = <String,Tag>{};

    await for( var tag in gitDir.tags()) {
      tags[tag.objectSha] = tag;
    }
```

The exception is located in tag.dart from git package:

```
  static Tag parseCatFile(String content) {
     ...
    final objectSha = headers['object']!.single;  <== "Null check operator used on a null value"
    final type = headers['type']!.single;
    final tag = headers['tag']!.single;
    final tagger = headers['tagger']!.single;

    return Tag._internal(objectSha, type, tag, tagger);
  }
```

Regards,
Marco

### PR DESCRIPTION
**Pull request by Stargator**
Resolves #31  and Thanks to @wgledbe built on top of #47.

Did not change @wgledbe's solution, just added tests to support the use cases.

@kevmoo, please let me know if you want anything else.

### CODE REVIEW

**Comment by kevmoo** [comment]
Looks like you need to rebase @Stargator !

File: lib/src/git_dir.dart [hunk 1]
```diff
@@ -69,7 +69,11 @@ class GitDir {
 
     for (var ref in refs) {
       final pr = await runCommand(['cat-file', '-p', ref.sha]);
-      yield Tag.parseCatFile(pr.stdout as String);
+      yield Tag.parseCatFile(
+        pr.stdout as String,
+        ref.sha,
+        ref.reference.substring(10),
```
Comment (kevmoo) [inline lib/src/git_dir.dart:75]: why always the substring here?

**Comment by Stargator** [comment]
@kevmoo, Yeah I did that earlier. But maybe I missed something. So I did another rebase

File: lib/src/git_dir.dart (hunk 1 above)
Comment (Stargator) [reply lib/src/git_dir.dart:75]: @wgledbe's changes, but I can look into it more and possible alternatives.

Comment (Stargator) [reply lib/src/git_dir.dart:75]: Looks like the tag's name starts at the 10th index, example of `ref.reference value`:

> refs/tags/newTag

**Comment by Stargator** [comment]
Fixed formatting issues from CI/analyze action

File: lib/src/git_dir.dart (hunk 1 above)
Comment (Stargator) [reply lib/src/git_dir.dart]: @kevmoo I replaced the substring with a split and last:

>   tag = ref.reference.split('/').last;

And moved it into the parsing method

File: lib/src/tag.dart [hunk 2]
```diff
@@ -32,10 +33,24 @@ class Tag {
...
+      // Annotated Tag
+      objectSha = headers['object']!.single;
+      type = headers['type']!.single;
+      tag = headers['tag']!.single;
+      tagger = headers['tagger']!.single;
+    } else {
+      // Lightweight Tag
+      objectSha = ref.sha;
+      type = 'lightweight';
+      tag = ref.reference.split('/').last;
```
Comment (kevmoo) [inline lib/src/tag.dart:51]: I think this should use a regex. And track the first bits stictly.

Comment (Stargator) [reply lib/src/tag.dart:51]: Just to clarify, you want the `split` to use a RegEx instead? Or to use something, like allMatches?

Is there a test case you are concerned about that I could implement so I know when I found a solution? RegEx is not my strong suit.

Comment (Stargator) [reply lib/src/tag.dart:51]: @kevmoo, I need more direction on what you want regarding the regex pattern (what's the targeted matches?) and clarity on what you mean by tracking first bits.

Comment (kevmoo) [reply lib/src/tag.dart]: I just pushed a commit!

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 2.3.0-wip
+
+- Added support for lightweight tags.
```
Comment (kevmoo) [inline CHANGELOG.md:3]: @Stargator – I did this for you. For future: pubspec and changelog entries!!

Comment (Stargator) [reply CHANGELOG.md:3]: Thanks, I also saw the changes to the test. I didn't know there was an hasLength matcher, so good to know.

This all looks good to me.

File: test/tag_test.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        '--message',
+        'First tag',
+        'anotherTag',
+        branchRef.sha,
+      ],
+      processWorkingDir: testDir.path,
+    );
+
+    final tagsFound = await testDir.tags().toList();
+    expect(tagsFound, hasLength(2));
```
Comment (kevmoo) [inline test/tag_test.dart:50]: This could use some more tests to make sure the actual tags content has what you expect @Stargator

Comment (Stargator) [reply test/tag_test.dart:50]: I can create an PR to do that, sure.

Comment (Stargator) [reply test/tag_test.dart:50]: Done #67

**Comment by wgledbe** [comment]
@Stargator Thanks for this!!

**Comment by Stargator** [comment]
@kevmoo when will a new version of the `git` package be released with this change?

**Comment by kevmoo** [comment]
> @kevmoo when will a new version of the `git` package be released with this change?

2.3.0 is now published!

**Comment by Stargator** [comment]
GREAT!

**Pull request merged**